### PR TITLE
New module vip-publication out of the existing publication in vip-core

### PR DESCRIPTION
**Pull request by camarasu**
Better modularity and the possibility to associate a VIPApplication to a publication (see Issue #220)

Changes/Improvements : 
- separate VIPPublication module
- Added a VIPApplication field in the Publication form and to VIPPublication DB (issue #220)
- VIPPublication DB is now created automatically if it does not exists
- VIPAuthor and VIPApplication are secondary keys, which allows automatic update (removal of the updateAuthor method)
- update of getApplications() to work for non-admin users also, by returning only the list of applications they have access to

### CODE REVIEW

File: vip-publication/src/main/java/fr/insalyon/creatis/vip/publication/client/view/Data.java [hunk 1]
```diff
@@ -29,15 +29,12 @@
  * The fact that you are presently reading this means that you have had
  * knowledge of the CeCILL-B license and that you accept its terms.
  */
-package fr.insalyon.creatis.vip.core.client.view.user.publication;
+package fr.insalyon.creatis.vip.publication.client.view;
```
Comment (axlbonnet) [inline vip-publication/src/main/java/fr/insalyon/creatis/vip/publication/client/view/Data.java:32]: Is it not necessary to add vipApplication in this class ? I'm lost with all those gwt grid data holder classes.

**Review by axlbonnet — CHANGES_REQUESTED** [review]
Good job to refactor this publication module ! Some small remarks about the gwt/java code.

Obviously the remaining job concerns database maintenance, in 2 steps : this PR and at release time.
I think we must, with this PR :
- rename the table to the original `VipPublication` in this PR code
- update the engine of the current `VipPublication` (myisam -> innodb)
- prepare the SQL requests to run on release to add the new column and the new constraints (to write in the `database_update.sql` at repository root)
- verify all the current publication vipAuthor actually exist to avoid a constraint issue when we will add the foreign key. 

At release time, we will only have to run the few SQL requests.

File: vip-application/src/main/java/fr/insalyon/creatis/vip/application/server/rpc/ApplicationServiceImpl.java [hunk 2]
```diff
@@ -225,8 +226,13 @@ public String publishVersion(String applicationName, String version) throws Appl
...
                 return applicationBusiness.getApplications(classes, connection);
             }
-            logger.error("Unauthorized to get all applications");
-            throw new ApplicationException("You have no administrator rights.");
+            List<AppClass> classes = classBusiness.getUserClasses(
+                    getSessionUser().getEmail(), false, connection);
+            List<String> classNames = new ArrayList<>();
+            for (AppClass c : classes) {
+                classNames.add(c.getName());
+            }
```
[the remark this answers is not in the document]
Comment (camarasu) [reply vip-application/src/main/java/fr/insalyon/creatis/vip/application/server/rpc/ApplicationServiceImpl.java]: Nice, thanks for the suggestion

File: vip-publication/src/main/java/fr/insalyon/creatis/vip/publication/client/view/EditPublicationLayout.java [hunk 3]
```diff
@@ -152,6 +161,7 @@ public void setPublication(String id, String title, String type, String nameType
             this.publicationDate.setValue("");
             this.doiField.setValue("");
             this.authorsField.setValue("");
+            this.vipApplication.setValue(vipApplication);
```
[the remark this answers is not in the document]
Comment (camarasu) [reply vip-publication/src/main/java/fr/insalyon/creatis/vip/publication/client/view/EditPublicationLayout.java]: Indeed.
I created another ticket to handle compulsory values

File: vip-publication/src/main/java/fr/insalyon/creatis/vip/publication/client/view/Data.java (hunk 1 above)
Comment (camarasu) [reply vip-publication/src/main/java/fr/insalyon/creatis/vip/publication/client/view/Data.java:32]: Done

**Comment by camarasu** [comment]
As discussed, there is no need to update VipPublication, neither to change myisam -> innodb, nor to add the new column and the new constraints, because I 
- created a new table (the code for the table creation is integrated in the PR)
- imported data from the old one
- renamed the new table with the old name (current VIP version works fine with the new table which has one extra column)

I also renamed the table to the original VipPublication in this PR code
So there should be no change required when getting the next release in production

**Comment by axlbonnet** [comment]
@camarasu  Doing tests, I found there is an issue with the new `PublicationParser testLastUpdatePublication` call.

The `callback` is `null` and I guess that GWT does not support its `final` field constuction.

**Pull request merged**